### ISSUE TITLE
binary_crossentropy provides wrong gradient

### ISSUE THREAD

**Issue by tangsanli5201** [keras-team/keras#9664]
Keras version: 2.0.8, backend tensorflow.
------------------------------------------------------------------------------------------------------------------
When I use binary_crossentropy as a loss function, I get wrong gradient when the logits is very large(with sigmoid activation). For example:
======================= example code ===================
import tensorflow as tf
from keras import metrics
import numpy as np

x = tf.placeholder(tf.float32, shape=1)   # logits
y = tf.placeholder(tf.float32, shape=1)   # label

sigmoid_x = tf.nn.sigmoid(x)
loss = metrics.binary_crossentropy(y, sigmoid_x)
gradients = tf.gradients(loss, sigmoid_x)

with tf.Session() as sess:
    grad = sess.run(gradients, feed_dict={
        x: np.array([100], np.float32),
        y: np.array([1], np.float32)
    })
======================== code end =======================
as calculated by hand, the output gradient (denoted as 'grad' in the code above) should be -1
however the output of keras is 0.
I think the problem is caused by the calculation precision.

**Comment by connection-on-fiber-bundles** [comment]
Could you also check the output of 
```
with tf.Session() as sess:
    print(sess.run(sigmoid_x, 
                             feed_dict={x: np.array([100], np.float32)}
                             )
            )
```
If `sigmoid_x` is actually evaluated as 1, which matches the expected `y`, then surely there's no error being back propagated.

**Comment by tangsanli5201** [comment]
@connection-on-fiber-bundles Yes. But the binary_crossentropy should act like this:
loss = - ( y*lnp + (1-y)*ln(1-p))
when y is equal to 1, the derivation to p should be 1/p, which means when p->1, the derivation should -> -1?
Actually, in the example code above, if we feed x=1, 10, 16, the output grad is -1.36787963, -1.00004542, -1.00000048 respectively( approaching to -1 ).

**Comment by connection-on-fiber-bundles** [comment]
@tangsanli5201 I get your points. That said, I believe (correct me if I'm wrong), the gradient here is for the parameters in the model, not for the `p` in your derivation. Let's say here
```
p = 1 / (1 + exp(-\beta x) 
```
then we should calculate the derivative of
```
-y \ln(1 / (1 + exp(-\beta x) ) - (1-y)\ln( exp(-\beta x) / (1 + exp(-\beta x) ) 
```
with respect to `\beta` instead (with y = 1 and x = 100).
Given that 
```
1 / (1 + exp(-\beta x) \approx 1
```
my calculation gives that the derivative is approximately just 0.

**Comment by connection-on-fiber-bundles** [comment]
BTW, for some intuition, your calculation basically says that we want to change `\beta` in the way that we can move `p` to the opposite of direction of -1, namely trying to increase p. But that's not possible since p is already 1 in this case. So I feel it's actually no surprising that we don't change `\beta` in this case (or that the back-propagated gradient is 0)

**Comment by tangsanli5201** [comment]
@connection-on-fiber-bundles 
I am sorry that I didn't describe my problem clearly.
For the binary cross-entropy: loss = - ( y*lnp + (1-y)*ln(1-p))
I do really need to calculate the gradient for  p not for the parameter (\beta as you illustrated) in the sigmoid function. Actually, the codes : 
                                              gradients = tf.gradients(loss, sigmoid_x)
calculate the gradient for sigmoid_x(which is 'p' in the equation, p=sigmoid_x=sigmoid(x)).
So, when x is large enough(e.g. x=100), p -> 1, the gradient should be -> -1.
As the toy experiments I illustrated above:
       " if we feed x=1, 10, 16, the output gradient is -1.36787963, -1.00004542, -1.00000048 respectively( approaching to -1".
      This shows that the gradient is exactly -> -1.

BTW, I notice that in Keras, it calls function 'sigmoid_cross_entropy_with_logits' in tensorflow backend in Keras function: keras.metrics.binary_crossentropy. However, in tensorflow backend, it calculate the binary_crossentropy from the logits (not the probabilities in keras). So in Keras, it first recovers the logits from the probabilities and then calls this tensorflow function. The Keras codes are:

def binary_crossentropy(target, output, from_logits=False):
    """Binary crossentropy between an output tensor and a target tensor.

    # Arguments
        target: A tensor with the same shape as `output`.
        output: A tensor.
        from_logits: Whether `output` is expected to be a logits tensor.
            By default, we consider that `output`
            encodes a probability distribution.

    # Returns
        A tensor.
    """
    # Note: tf.nn.sigmoid_cross_entropy_with_logits
    # expects logits, Keras expects probabilities.
    if not from_logits:
        # transform back to logits
        _epsilon = _to_tensor(epsilon(), output.dtype.base_dtype)
        output = tf.clip_by_value(output, _epsilon, 1 - _epsilon)
        output = tf.log(output / (1 - output))

    return tf.nn.sigmoid_cross_entropy_with_logits(labels=target,
                                                   logits=output)

So, I wonder whether it is the epsilon causing such 'wrong gradient' problem?

**Comment by connection-on-fiber-bundles** [comment]
@tangsanli5201 I see what you mean now. Sorry for the previous misunderstanding :)

Hmm yeah the results from your toy experiments do indicate something might have gone wrong when x is 100. I don't really know which part goes wrong though
